src/components/ui/calendar.tsx

```diff
@@ -17,111 +16,118 @@ import {
   SelectTrigger,
 } from "@/components/ui/select";
 
-import { setMonth } from "@/lib/config/date-fns";
+import { addMonths } from "@/lib/config/date-fns";
 
 import { cn } from "@/lib/utils";
 
 export type CalendarProps = React.ComponentProps<typeof DayPicker>;
 
-const DEFAULT_FROM_YEAR = 1900;
-
 function Calendar({
   className,
   classNames,
   showOutsideDays = true,
-  fromYear,
-  toYear,
+  startMonth,
+  endMonth,
+  hidden,
   captionLayout,
   ...props
 }: CalendarProps) {
-  const layout: CaptionLayout = captionLayout || "dropdown-buttons";
+  const layout: CalendarProps["captionLayout"] = captionLayout || "dropdown";
+
+  const calendarStartMonth = startMonth ?? new Date(1900, 0, 0);
+  const calendarEndMonth =
+    endMonth ?? new Date(new Date().getFullYear() + 15, 11);
 
-  const currentYear = new Date().getFullYear();
+  const hiddenIsArray = Array.isArray(hidden);
 
-  const oldestYear = fromYear ?? DEFAULT_FROM_YEAR;
-  const newestYear = toYear ?? currentYear + 15;
+  const calendarHiddenOptions: Matcher[] = [
```

Comment:
Relying on the react-compiler to perform auto memoizations here for better perf.